pkg/ssh/keygen.go

```diff
@@ -187,17 +192,39 @@ type PublicKey struct {
 // ExtractPublicKey extracts and returns the public key from the specified
 // private key, along with its fingerprint hashes.
 func ExtractPublicKey(privateKeyPath string) (PublicKey, error) {
-	keyBytes, err := exec.Command("ssh-keygen", "-y", "-f", privateKeyPath).CombinedOutput()
+	// Try to read public key from disk
+	publicKeyFilePath := privateKeyPath + ".pub"
+	keyPathForFingerprints := publicKeyFilePath
+
+	keyBytes, err := ioutil.ReadFile(publicKeyFilePath)
 	if err != nil {
-		return PublicKey{}, errors.New(string(keyBytes))
+		// Public file doesn't exist, or isn't readable
+		// Generate it from the private key instead
+		fileInfo, err := os.Stat(privateKeyPath)
+
+		if fileInfo.Mode() != PrivateKeyFileMode {
+			// The key is mounted, but not the right permissions; since
+			// it's likely to be read-only, we may not be able to rectify
+			// this, but let's try.
+			if err := os.Chmod(privateKeyPath, PrivateKeyFileMode); err != nil {
+				return PublicKey{}, errors.Wrap(err, "failed to chmod identity file")
+			}
+		}
+
+		keyBytes, err = exec.Command("ssh-keygen", "-y", "-f", privateKeyPath).CombinedOutput()
+		if err != nil {
+			return PublicKey{}, errors.New(string(keyBytes))
+		}
+
+		keyPathForFingerprints = privateKeyPath
```

Comment:
This is the previous public key generating logic which is invoked if the public key file does not already exist